### ISSUE TITLE
Machine stopped unexpectedly inside fork/join

### ISSUE THREAD

**Issue by netalpha** [spring-projects/spring-statemachine#737]
SM version: 2.1.2.RELEASE

Sample code: https://github.com/netalpha/spring-state-machine-example

The state machine diagram:
![state-machine-sample](https://user-images.githubusercontent.com/3212787/57070988-fd4a1980-6cd0-11e9-8cd6-7083b923b9af.png)

When the `NEW` event is sent, I am expecting the SM context to be: [S2, S21I, S22I, S23_IN_PROGRESS, S24E]. 
However,  the actual result is  [S2, S21I, S22I, S23_IN_PROGRESS, S24I]. 

The logs as follows:
```
2019-05-02 11:08:55.412  INFO 65567 --- [nio-8080-exec-2] o.s.s.s.DefaultStateMachineService       : Acquiring machine with id datajpapersist1
2019-05-02 11:08:55.413  INFO 65567 --- [nio-8080-exec-2] o.s.s.s.DefaultStateMachineService       : Getting new machine from factory with id datajpapersist1
2019-05-02 11:08:55.558  INFO 65567 --- [nio-8080-exec-2] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@24677b6c
2019-05-02 11:08:55.559  INFO 65567 --- [nio-8080-exec-2] o.s.s.support.LifecycleObjectSupport     : started S0 S3 S5 S1 S4 S2 S22E S22I S22_IN_PROGRESS S22_PASSED S21I S21E S21_NOT_REQUIRED S21_IN_PROGRESS S24_PASSED S24_NOT_REQUIRED S24I S24E S24_IN_PROGRESS CHOICE_S24 S23E S23I S23_IN_PROGRESS S23_PASSED FORK_S2 JOIN_S2  / S0 / uuid=b22a80ab-9349-4d4b-9e16-ed4ab5c56758 / id=datajpapersist1
2019-05-02 11:08:55.559  INFO 65567 --- [nio-8080-exec-2] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@114bd1ab
2019-05-02 11:08:55.559  INFO 65567 --- [nio-8080-exec-2] o.s.s.support.LifecycleObjectSupport     : started ROOT S0 S3 S5 S1 S4 S2 S22E S22I S22_IN_PROGRESS S22_PASSED S21I S21E S21_NOT_REQUIRED S21_IN_PROGRESS S24_PASSED S24_NOT_REQUIRED S24I S24E S24_IN_PROGRESS CHOICE_S24 S23E S23I S23_IN_PROGRESS S23_PASSED FORK_S2 JOIN_S2  / ROOT,S0 / uuid=4bb6dab5-d3dd-49f9-ae49-66cbaf8c69f5 / id=datajpapersist1
2019-05-02 11:08:58.780  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : stopped org.springframework.statemachine.support.DefaultStateMachineExecutor@24677b6c
2019-05-02 11:08:58.783  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : stopped S0 S3 S5 S1 S4 S2 S22E S22I S22_IN_PROGRESS S22_PASSED S21I S21E S21_NOT_REQUIRED S21_IN_PROGRESS S24_PASSED S24_NOT_REQUIRED S24I S24E S24_IN_PROGRESS CHOICE_S24 S23E S23I S23_IN_PROGRESS S23_PASSED FORK_S2 JOIN_S2  /  / uuid=b22a80ab-9349-4d4b-9e16-ed4ab5c56758 / id=datajpapersist1
2019-05-02 11:08:58.790  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@4adb8dbc
2019-05-02 11:08:58.791  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started S22E S22I S22_IN_PROGRESS S22_PASSED  / S22I / uuid=76b7e66d-e0ca-49fb-abb4-c9cdfb545795 / id=datajpapersist1#062688b5-98b2-4d2d-bcfa-55baf9900f6d
2019-05-02 11:08:58.793  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@8b6ee2f
2019-05-02 11:08:58.794  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started S21I S21E S21_NOT_REQUIRED S21_IN_PROGRESS  / S21I / uuid=afc845b0-2382-4200-8c8c-1b30a09e2315 / id=datajpapersist1#c67d5e7f-528b-4a6a-8807-9b2de66339b1
s24-choice
s24-not-required
2019-05-02 11:08:58.803  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : stopped org.springframework.statemachine.support.DefaultStateMachineExecutor@4adb8dbc
2019-05-02 11:08:58.804  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : stopped S22E S22I S22_IN_PROGRESS S22_PASSED  /  / uuid=76b7e66d-e0ca-49fb-abb4-c9cdfb545795 / id=datajpapersist1#062688b5-98b2-4d2d-bcfa-55baf9900f6d
2019-05-02 11:08:58.804  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : stopped org.springframework.statemachine.support.DefaultStateMachineExecutor@8b6ee2f
2019-05-02 11:08:58.805  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : stopped S21I S21E S21_NOT_REQUIRED S21_IN_PROGRESS  /  / uuid=afc845b0-2382-4200-8c8c-1b30a09e2315 / id=datajpapersist1#c67d5e7f-528b-4a6a-8807-9b2de66339b1
2019-05-02 11:08:58.805  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : stopped org.springframework.statemachine.support.DefaultStateMachineExecutor@2e654faf
2019-05-02 11:08:58.805  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : stopped S24_PASSED S24_NOT_REQUIRED S24I S24E S24_IN_PROGRESS CHOICE_S24  /  / uuid=02c8289d-6028-4ea4-a2df-c5c071a1a78d / id=datajpapersist1#e7f90d11-08d2-4cdc-a2d3-7aa9c217935c
2019-05-02 11:08:58.809  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@2e654faf
2019-05-02 11:08:58.810  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started S24_PASSED S24_NOT_REQUIRED S24I S24E S24_IN_PROGRESS CHOICE_S24  / S24E / uuid=02c8289d-6028-4ea4-a2df-c5c071a1a78d / id=datajpapersist1#e7f90d11-08d2-4cdc-a2d3-7aa9c217935c
s23-inprogress
2019-05-02 11:08:58.817  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@16bd6e24
2019-05-02 11:08:58.817  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started S23E S23I S23_IN_PROGRESS S23_PASSED  / S23_IN_PROGRESS / uuid=fbe39145-3bf5-48c0-93af-edc190b1d145 / id=datajpapersist1#de13a92a-10f8-44dc-bc39-1473923779ec
2019-05-02 11:08:58.818  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@24677b6c
2019-05-02 11:08:58.820  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@4adb8dbc
2019-05-02 11:08:58.820  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started S22E S22I S22_IN_PROGRESS S22_PASSED  / S22I / uuid=76b7e66d-e0ca-49fb-abb4-c9cdfb545795 / id=datajpapersist1#062688b5-98b2-4d2d-bcfa-55baf9900f6d
2019-05-02 11:08:58.821  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@8b6ee2f
2019-05-02 11:08:58.822  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started S21I S21E S21_NOT_REQUIRED S21_IN_PROGRESS  / S21I / uuid=afc845b0-2382-4200-8c8c-1b30a09e2315 / id=datajpapersist1#c67d5e7f-528b-4a6a-8807-9b2de66339b1
2019-05-02 11:08:58.822  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@2e654faf
2019-05-02 11:08:58.822  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started S24_PASSED S24_NOT_REQUIRED S24I S24E S24_IN_PROGRESS CHOICE_S24  / S24E / uuid=02c8289d-6028-4ea4-a2df-c5c071a1a78d / id=datajpapersist1#e7f90d11-08d2-4cdc-a2d3-7aa9c217935c
2019-05-02 11:08:58.822  INFO 65567 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started S0 S3 S5 S1 S4 S2 S22E S22I S22_IN_PROGRESS S22_PASSED S21I S21E S21_NOT_REQUIRED S21_IN_PROGRESS S24_PASSED S24_NOT_REQUIRED S24I S24E S24_IN_PROGRESS CHOICE_S24 S23E S23I S23_IN_PROGRESS S23_PASSED FORK_S2 JOIN_S2  / S2,S22I,S21I,S24I,S23_IN_PROGRESS / uuid=b22a80ab-9349-4d4b-9e16-ed4ab5c56758 / id=datajpapersist1

```

What I can see from the log is that the state machine for region `S21` `S22` `S24` stopped and then start again, but `S23` did not because it was transiting to `S23_IN_PROGRESS`. 
I think this expectedly stopping caused `S24` did not go to `S24E` in the end.

My understanding of this issue is none of `S21`. `S22`. `S23` `S24` should be stopped at this stage. 

I tried to comment line: 1178 in `AbstractStateMachine.class`. 
```
void exitCurrentState(State<S, E> state, Message<E> message, Transition<S, E> transition, StateMachine<S, E> stateMachine,
			Collection<State<S, E>> sources, Collection<State<S, E>> targets) {
		if (currentState == null) {
			return;
		}
		if (currentState.isSubmachineState()) {
			StateMachine<S, E> submachine = ((AbstractState<S, E>)currentState).getSubmachine();
			((AbstractStateMachine<S, E>)submachine).exitCurrentState(state, message, transition, stateMachine);
			exitFromState(currentState, message, transition, stateMachine, sources, targets);
		} else if (currentState.isOrthogonal()) {
			Collection<Region<S,E>> regions = ((AbstractState<S, E>)currentState).getRegions();
			for (Region<S,E> r : regions) {
				if (r.getStates().contains(state)) {
					exitFromState(r.getState(), message, transition, stateMachine, sources, targets);
				}
			}
//			exitFromState(currentState, message, transition, stateMachine, sources, targets);
		} else {
			exitFromState(currentState, message, transition, stateMachine, sources, targets);
		}
	}
```

The state machine gave me the correct final context [S2,S21I,S22I,S23_IN_PROGRESS,S24E] as shown in the following logs, but the S24 region was still stopped and started twice.

```
2019-05-02 12:17:27.195  INFO 70291 --- [nio-8080-exec-2] o.s.s.s.DefaultStateMachineService       : Acquiring machine with id datajpapersist1
2019-05-02 12:17:27.195  INFO 70291 --- [nio-8080-exec-2] o.s.s.s.DefaultStateMachineService       : Getting new machine from factory with id datajpapersist1
2019-05-02 12:17:27.373  INFO 70291 --- [nio-8080-exec-2] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@1d500b5b
2019-05-02 12:17:27.374  INFO 70291 --- [nio-8080-exec-2] o.s.s.support.LifecycleObjectSupport     : started S3 S2 S21I S21_IN_PROGRESS S21E S21_NOT_REQUIRED S22_PASSED S22E S22_IN_PROGRESS S22I S23I S23_IN_PROGRESS S23E S23_PASSED S24I S24_PASSED S24_NOT_REQUIRED S24E S24_IN_PROGRESS CHOICE_S24 S5 S0 S1 S4 JOIN_S2 FORK_S2  / S0 / uuid=521bf8e1-067e-4da2-8877-a21d1e4fbbb0 / id=datajpapersist1
2019-05-02 12:17:27.374  INFO 70291 --- [nio-8080-exec-2] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@62a521b7
2019-05-02 12:17:27.374  INFO 70291 --- [nio-8080-exec-2] o.s.s.support.LifecycleObjectSupport     : started ROOT S3 S2 S21I S21_IN_PROGRESS S21E S21_NOT_REQUIRED S22_PASSED S22E S22_IN_PROGRESS S22I S23I S23_IN_PROGRESS S23E S23_PASSED S24I S24_PASSED S24_NOT_REQUIRED S24E S24_IN_PROGRESS CHOICE_S24 S5 S0 S1 S4 JOIN_S2 FORK_S2  / ROOT,S0 / uuid=715bc676-0026-403a-a9b2-115c90083461 / id=datajpapersist1
2019-05-02 12:17:32.205  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : stopped org.springframework.statemachine.support.DefaultStateMachineExecutor@1d500b5b
2019-05-02 12:17:32.209  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : stopped S3 S2 S21I S21_IN_PROGRESS S21E S21_NOT_REQUIRED S22_PASSED S22E S22_IN_PROGRESS S22I S23I S23_IN_PROGRESS S23E S23_PASSED S24I S24_PASSED S24_NOT_REQUIRED S24E S24_IN_PROGRESS CHOICE_S24 S5 S0 S1 S4 JOIN_S2 FORK_S2  /  / uuid=521bf8e1-067e-4da2-8877-a21d1e4fbbb0 / id=datajpapersist1
2019-05-02 12:17:32.218  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@64253fde
2019-05-02 12:17:32.219  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started S21I S21_IN_PROGRESS S21E S21_NOT_REQUIRED  / S21I / uuid=33042c41-a903-40fb-96d5-c629e68bbd72 / id=datajpapersist1#ea3bec18-5e6b-4391-a17b-1dc72da941ad
2019-05-02 12:17:32.224  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@59c54d82
2019-05-02 12:17:32.225  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started S22_PASSED S22E S22_IN_PROGRESS S22I  / S22I / uuid=c2982b7a-8bd7-4a66-a4a3-7223109dc2b6 / id=datajpapersist1#98b7202a-5553-49ae-b2cf-87e631cd86a8
s23-inprogress
2019-05-02 12:17:32.235  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@2dc3c2e2
2019-05-02 12:17:32.235  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started S23I S23_IN_PROGRESS S23E S23_PASSED  / S23_IN_PROGRESS / uuid=36d11bb0-e290-462a-a0aa-a806bf38c013 / id=datajpapersist1#c3976832-0f7d-4c28-bcdb-2873401615c3
s24-choice
s24-not-required
2019-05-02 12:17:32.248  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : stopped org.springframework.statemachine.support.DefaultStateMachineExecutor@6ff56d89
2019-05-02 12:17:32.248  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : stopped S24I S24_PASSED S24_NOT_REQUIRED S24E S24_IN_PROGRESS CHOICE_S24  /  / uuid=174fea16-44b8-41da-be96-87024ae606af / id=datajpapersist1#6061d623-4b21-402f-afdd-cfffad748ad0
2019-05-02 12:17:32.249  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@6ff56d89
2019-05-02 12:17:32.249  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started S24I S24_PASSED S24_NOT_REQUIRED S24E S24_IN_PROGRESS CHOICE_S24  /  / uuid=174fea16-44b8-41da-be96-87024ae606af / id=datajpapersist1#6061d623-4b21-402f-afdd-cfffad748ad0
2019-05-02 12:17:32.249  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@1d500b5b
s24-choice
s24-not-required
2019-05-02 12:17:32.258  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : stopped org.springframework.statemachine.support.DefaultStateMachineExecutor@6ff56d89
2019-05-02 12:17:32.258  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : stopped S24I S24_PASSED S24_NOT_REQUIRED S24E S24_IN_PROGRESS CHOICE_S24  /  / uuid=174fea16-44b8-41da-be96-87024ae606af / id=datajpapersist1#6061d623-4b21-402f-afdd-cfffad748ad0
2019-05-02 12:17:32.259  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started org.springframework.statemachine.support.DefaultStateMachineExecutor@6ff56d89
2019-05-02 12:17:32.259  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started S24I S24_PASSED S24_NOT_REQUIRED S24E S24_IN_PROGRESS CHOICE_S24  /  / uuid=174fea16-44b8-41da-be96-87024ae606af / id=datajpapersist1#6061d623-4b21-402f-afdd-cfffad748ad0
2019-05-02 12:17:32.259  INFO 70291 --- [nio-8080-exec-3] o.s.s.support.LifecycleObjectSupport     : started S3 S2 S21I S21_IN_PROGRESS S21E S21_NOT_REQUIRED S22_PASSED S22E S22_IN_PROGRESS S22I S23I S23_IN_PROGRESS S23E S23_PASSED S24I S24_PASSED S24_NOT_REQUIRED S24E S24_IN_PROGRESS CHOICE_S24 S5 S0 S1 S4 JOIN_S2 FORK_S2  / S2,S21I,S22I,S23_IN_PROGRESS,S24E / uuid=521bf8e1-067e-4da2-8877-a21d1e4fbbb0 / id=datajpapersist1
```

Any idea?

**Comment by jvalkeal** [comment]
Your PR breaks a lot of tests but thanks for a sample. You're right, it should go to `S24E`. I'll try to see how it runs.

**Comment by jvalkeal** [comment]
Right, machine actually goes to `S24E` but it's reported wrongly from:

https://github.com/spring-projects/spring-statemachine/blob/dc72cc4982126542827616e44c0b56f957314b9b/spring-statemachine-core/src/main/java/org/springframework/statemachine/support/AbstractStateMachine.java#L182-L191

When a region goes to a terminal/end state it is finished and that's why region is stopped. I'll poke around a bit more and create a new issue before I close this one.

**Comment by netalpha** [comment]
I agree S24 region would stop, but how about S21, S22. S21 and S22 are both at the initial stage, but they stopped as well.

**Comment by jvalkeal** [comment]
Indeed we probably have two issues here. Looks like all those 4 regions inside `S2` are running but `S2` is not. I'll try to transfer this machine as a test to a framework to figure out where things go wrong.

**Comment by jvalkeal** [comment]
@netalpha Hopefully fixed now in master. It was a good case to point out as we didn't have a test for this particular scenario. Lemmy know how it goes.

**Comment by hamid646** [comment]
any chance this could be released  today (i work with @netalpha )?

**Comment by netalpha** [comment]
> @netalpha Hopefully fixed now in master. It was a good case to point out as we didn't have a test for this particular scenario. Lemmy know how it goes.

Yes, it works. Thanks for the fix.

**Comment by hamid646** [comment]
@jvalkeal could you please release this fix for 2.1.x? we really need this fix

**Comment by hamid646** [comment]
@jvalkeal any updates ?
